src/contracts/atlas/Sorter.sol

```diff
@@ -0,0 +1,199 @@
+        // TODO: subtract any pending withdrawals
+        uint256 searcherBalance = IEscrow(escrow).searcherEscrowBalance(searcherCall.metaTx.from);
+        if (searcherBalance < searcherCall.metaTx.maxFeePerGas * searcherCall.metaTx.gas) {
+            return false;
+        }
+
+        // Searchers can only do one tx per block - this prevents double counting escrow balances.
+        // TODO: Add in "targetBlockNumber" as an arg?
+        uint256 searcherLastActiveBlock = IEscrow(escrow).searcherLastActiveBlock(searcherCall.metaTx.from);
+        if (searcherLastActiveBlock >= block.number) {
+            return false;
+        }
+
+        // Make sure the searcher nonce is accurate
+        uint256 nextSearcherNonce = IEscrow(escrow).nextSearcherNonce(searcherCall.metaTx.from);
+        if (nextSearcherNonce != searcherCall.metaTx.nonce) {
+            return false;
+        }
+
+        // Make sure that the searcher has the correct codehash for protocol control contract
+        if (protocolCall.to.codehash != searcherCall.metaTx.controlCodeHash) {
+            return false;
+        }
+
+        // Make sure that the searcher's maxFeePerGas matches or exceeds the user's
+        if (searcherCall.metaTx.maxFeePerGas < userMetaTx.maxFeePerGas) {
+            return false;
+        }
+
+        return true;
+    }
+
+    function _getSortingData(
+        ProtocolCall memory protocolCall, 
+        UserCall calldata userCall, 
+        SearcherCall[] calldata searcherCalls,
+        uint256 count
+    ) internal view returns (SortingData[] memory, uint256){
+
+        BidData[] memory bidFormat = IProtocolControl(protocolCall.to).getBidFormat(userCall.metaTx);
+
+        SortingData[] memory sortingData = new SortingData[](count);
+
+        uint256 i;
+        uint256 invalid;
+        for (;i<count;) {
+            if (
+                _verifyBidFormat(bidFormat, searcherCalls[i]) && 
+                _verifySearcherEligibility(protocolCall, userCall.metaTx, searcherCalls[i])
+            ) {
+                sortingData[i] = SortingData({
+                    amount: IProtocolControl(protocolCall.to).getBidValue(searcherCalls[i]),
+                    valid: true
+                });
+                
+
+            } else {
+                sortingData[i] = SortingData({
+                    amount: 0,
+                    valid: false
+                });
+                unchecked{ ++invalid; }
+            }
+            unchecked{ ++i; }            
+        }
+
+        return (sortingData, invalid);
+    }
+
+    function _sort(
+        SortingData[] memory sortingData,
+        uint256 count,
+        uint256 invalid
+    ) internal pure returns (uint256[] memory) {
+
+        uint256[] memory sorted = new uint256[](count - invalid);
+
+        uint256 n; // outer loop counter
+        uint256 i; // inner loop counter
+
+        uint256 topBid;
+        uint256 bottomBid;
+
+        for (;invalid<count;) {
```

Comment:
We could explore a more efficient sorting algo here depending on the expected sizes of sortingData arrays.

Reply:
From 5/9 tweet in 2022: Quick-sort + insertion sort when partitions' length <= 7 
https://twitter.com/fiveoutofnine/status/1502988952428556291

Reply:
Let's explore it. I'm 100% on board with using a more efficient sorting algorithm.